server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -103,51 +125,80 @@ void setEnabled(boolean enabled) {
+                // reported health info gets reset to null.
+                lastReportedDiskHealthInfo.set(null);
+                healthNodeSelected = current != null;
+                if (healthNodeSelected) {
+                    maybeScheduleNow();
+                }
+            }
         }
     }
 
-    // Visible for testing
-    void monitorHealth() {
-        if (inProgress.compareAndSet(false, true)) {
+    private void monitorHealth() {
+        if (inProgress.compareAndSet(false, true) && healthNodeSelected) {
             ClusterState clusterState = clusterService.state();
             HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
             assert healthMetadata != null : "health metadata should have been initialized.";
-            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo;
+            DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo.get();
             DiskHealthInfo currentHealth = diskCheck.getHealth(healthMetadata, clusterState);
             if (currentHealth.equals(previousHealth) == false) {
-                logger.debug("Health status changed from {} to {}", previousHealth, currentHealth);
-                this.lastReportedDiskHealthInfo = currentHealth;
+                String nodeId = clusterService.localNode().getId();
+                ActionListener<AcknowledgedResponse> listener = ActionListener.wrap(response -> {
+                    // Update the last reported value if there was no change, that would mean that there
+                    // is a new health node that might have not received the value, so we should retry.
+                    lastReportedDiskHealthInfo.compareAndSet(previousHealth, currentHealth);
+                    logger.debug(
+                        "Health info [{}] successfully send, last reported value: {}.",
+                        currentHealth,
+                        lastReportedDiskHealthInfo.get()
+                    );
+                }, e -> logger.error(() -> format("Failed to send health info [%s] to health node, will try again.", currentHealth), e));
+                client.execute(
+                    UpdateHealthInfoCacheAction.INSTANCE,
+                    new UpdateHealthInfoCacheAction.Request(nodeId, currentHealth),
+                    ActionListener.runAfter(listener, () -> {
+                        inProgress.set(false);
+                        // Scheduling happens after the flag inProgress is false, this ensures that
+                        // if the feature is enabled after the following schedule statement, the setEnabled
+                        // method will be able to schedule the next run, and it will not be a no-op.
+                        // We prefer to err towards an extra scheduling than miss the enabling of this feature alltogether.
+                        maybeScheduleNextRun(monitorInterval);
```

Comment:
We only schedule a next run if currentHealth is not equal to previousHealth. I assume we want to schedule it regardless right?

Reply:
And related, we'll want to set inProgress to false even if we don't run `UpdateHealthInfoCacheAction`.

Reply:
Definitely, I completely missed it!